### PR TITLE
Update the documentation around the gocounter examples

### PR DESCRIPTION
**Pull request by Billy99**
Update the documentation around the gocounter examples to be inline with the addition of bpfd-operator.

Signedby: Billy McFall <22157057+Billy99@users.noreply.github.com>

### CODE REVIEW

File: docs/developer/example-bpf.md [hunk 1]
```diff
@@ -0,0 +1,670 @@
+# Example BPF Programs
+
+Example applications that uses the `bpfd-go` bindings can be found in
```
Comment (astoycos) [inline docs/developer/example-bpf.md:3]:
```suggestion
Example applications that use the `bpfd-go` bindings can be found in
```

File: docs/developer/example-bpf.md [hunk 2]
```diff
@@ -0,0 +1,670 @@
...
+[examples/go-xdp-counter/](https://github.com/redhat-et/bpfd/tree/main/examples/go-xdp-counter) and
+[examples/go-tc-counter/](https://github.com/redhat-et/bpfd/tree/main/examples/go-tc-counter).
+These examples and the associated documentation is intended to provide the basics on deploying a BPF
```
Comment (astoycos) [inline docs/developer/example-bpf.md:6]:
```suggestion
These examples and the associated documentation is intended to provide the basics on how to deploy and manage a BPF program using bpfd.
```

File: docs/developer/example-bpf.md [hunk 3]
```diff
@@ -0,0 +1,670 @@
...
+a bpfd-client certificate is used that is generated by bpfd.
+The diagram below shows `go-xdp-counter` example, but the `go-tc-counter` example
+operates exactly the same way.
+
+![go-xdp-counter On Host](../img/gocounter-on-host.png)
+
+Following the diagram (Purple numbers):
+1. When `go-xdp-counter` userspace is started, it will send a gRPC request
+over mTLS to `bpfd` requesting `bpfd` to load the `go-xdp-counter` BPF bytecode located at
+`src/bpfd/src/examples/go-xdp-counter/bpf_bpfel.o` at a priority of 50 and on interface `ens3`.
```
Comment (astoycos) [inline docs/developer/example-bpf.md:111]:
```suggestion
`bpfd/examples/go-xdp-counter/bpf_bpfel.o` at a priority of 50 and on interface `ens3`.
```

File: docs/developer/example-bpf.md [hunk 4]
```diff
@@ -0,0 +1,670 @@
...
+These values are configurable as we will see later, but for now we will use the defaults
+(except interface, which is required to be entered).
+2. `bpfd` will load it's `dispatcher` BPF program, which links to the `go-xdp-counter` BPF program
```
Comment (astoycos) [inline docs/developer/example-bpf.md:114]:
```suggestion
2. `bpfd` will load it's `dispatcher` BPF program, which links to the `go-xdp-counter` BPF program and return a UUID describing the running program.
```

File: docs/developer/example-bpf.md [hunk 5]
```diff
@@ -0,0 +1,670 @@
...
+
+#### Running Unprivileged
+
+To run the `go-xdp-counter` and `go-tc-counter` examples unprivileged (without `sudo`), the
+following three steps must be performed.
+
+##### Step 1: Create `bpfd` User Group
+
+The [tutorial.md](../admin/tutorial.md) guide describes the different modes `bpfd` can be run in.
+Specifically, the [Systemd Service](../admin/tutorial.md#systemd-service) section describe how to
```
Comment (astoycos) [inline docs/developer/example-bpf.md:187]:
```suggestion
Specifically, the [Systemd Service](../admin/tutorial.md#systemd-service) section explains how to
```

File: docs/developer/example-bpf.md [hunk 6]
```diff
@@ -0,0 +1,670 @@
...
+start `bpfd` the `bpfd` Users and `bpfd` User Group.
+`bpfd` must be started as a Systemd Service and `go-xdp-counter` must be run from a User that is a
```
Comment (astoycos) [inline docs/developer/example-bpf.md:189]: All this about the shared Socket is stale right? We got rid of passing the map FD over a socket

File: docs/developer/example-bpf.md [hunk 7]
```diff
@@ -0,0 +1,670 @@
...
+program.
+The approach is slightly different when running on a Kubernetes cluster.
+The BPF bytecode should be loaded by an administrator, not the userspace program itself.
+
+![gocounter On Kubernetes](../img/gocounter-on-k8s.png)
+
+### Loading BPF Bytecode On Kubernetes
+
+This assumes there is already a Kubernetes cluster running and `bpfd` is running in the cluster
+(see [How to Manually Deploy bpfd on Kubernetes](../admin/k8s-deployment.md)).
```
Comment (astoycos) [inline docs/developer/example-bpf.md:390]: This like points to a super stake doc currently.... (I'm assuming what @anfredette Is doing will refactor it?)

**Review by astoycos — CHANGES_REQUESTED** [review]
Just some comments/nits thanks @Billy99

File: docs/developer/example-bpf.md [hunk 8]
```diff
@@ -0,0 +1,670 @@
...
+
+    kubectl delete -f bpfd-operator/config/rbac/ebpfprogram_viewer_role.yaml
+     clusterrole.rbac.authorization.k8s.io "bpfprogram-viewer-role" deleted
+```
+
+## Notes
+
+Notes regarding this document:
+* Source of images used in this document can be found in
+  [bpfd Upstream Images](https://docs.google.com/presentation/d/1wU4xu6xeyk9cB3G-Nn-dzkf90j1-EI4PB167G7v-Xl4/edit?usp=sharing).
```
[the remark this answers is not in the document]
Comment (Billy99) [reply docs/developer/example-bpf.md:669]: No, worked for me. It's on my Red Hat account, so need to be logged in as Red Hat.

File: docs/developer/example-bpf.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (Billy99) [reply docs/developer/example-bpf.md:3]: Done

File: docs/developer/example-bpf.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (Billy99) [reply docs/developer/example-bpf.md:6]: Done

File: docs/developer/example-bpf.md [hunk 9]
```diff
@@ -0,0 +1,670 @@
+# Example BPF Programs
+
+Example applications that uses the `bpfd-go` bindings can be found in
+[examples/go-xdp-counter/](https://github.com/redhat-et/bpfd/tree/main/examples/go-xdp-counter) and
+[examples/go-tc-counter/](https://github.com/redhat-et/bpfd/tree/main/examples/go-tc-counter).
+These examples and the associated documentation is intended to provide the basics on deploying a BPF
+program using bpfd with a go gRPC client.
+There are two ways to deploy these example applications, simply run locally, or run in a Kubernetes
```
[the remark this answers is not in the document]
Comment (Billy99) [reply docs/developer/example-bpf.md:8]: Done

File: docs/developer/example-bpf.md [hunk 10]
```diff
@@ -0,0 +1,670 @@
...
+cluster.
+
+The `go-xdp-counter` and  `go-tc-counter` examples each contain a BPF Program written in C
+([xdp_counter.c](https://github.com/redhat-et/bpfd/tree/main/examples/go-xdp-counter/bpf/xdp_counter.c) and
+[tc_counter.c](https://github.com/redhat-et/bpfd/tree/main/examples/go-tc-counter/bpf/tc_counter.c))
+that is compiled into BPF bytecode.
+The BPF program counts packets that are received on an interface and stores the packet and byte
+counts in a map that is accessible by the userspace portion.
+The `go-xdp-counter` and  `go-tc-counter` examples each also have a userspace portion written in GO that is
```
[the remark this answers is not in the document]
Comment (Billy99) [reply docs/developer/example-bpf.md:17]: Rewritten

File: docs/developer/example-bpf.md [hunk 11]
```diff
@@ -0,0 +1,670 @@
...
+responsible for loading the BPF program at the XDP hook point or the TC hook point, via gRPC calls to `bpfd`, and
+then every 3 seconds polls the BPF map and logs the current counts.
+The userspace code is leveraging the [cilium/ebpf library](https://github.com/cilium/ebpf)
+to manage the maps shared with the BPF program.
+The programs are very similar, just varies where in the Linux networking stack the programs are inserted.
```
[the remark this answers is not in the document]
Comment (Billy99) [reply docs/developer/example-bpf.md:22]: Done

File: docs/developer/example-bpf.md [hunk 12]
```diff
@@ -0,0 +1,670 @@
...
+    go build
+```
+
+### Running On Host
+
+The most basic way to deploy this example is running directly on a host system.
+First, start or ensure `bpfd` is up and running.
+[tutorial.md](../admin/tutorial.md) will guide you through deploying `bpfd`.
+In all the examples of running `go-xdp-counter` or `go-tc-counter` on a host system,
+a bpfd-client certificate is used that is generated by bpfd.
```
[the remark this answers is not in the document]
Comment (Billy99) [reply docs/developer/example-bpf.md:102]: Done

File: docs/developer/example-bpf.md [hunk 13]
```diff
@@ -0,0 +1,670 @@
...
+The diagram below shows `go-xdp-counter` example, but the `go-tc-counter` example
+operates exactly the same way.
+
+![go-xdp-counter On Host](../img/gocounter-on-host.png)
+
+Following the diagram (Purple numbers):
+1. When `go-xdp-counter` userspace is started, it will send a gRPC request
+over mTLS to `bpfd` requesting `bpfd` to load the `go-xdp-counter` BPF bytecode located at
```
[the remark this answers is not in the document]
Comment (Billy99) [reply docs/developer/example-bpf.md:110]: Done

File: docs/developer/example-bpf.md (hunk 3 above)
Comment (Billy99) [reply docs/developer/example-bpf.md:111]: Good catch. Done

File: docs/developer/example-bpf.md (hunk 4 above)
Comment (Billy99) [reply docs/developer/example-bpf.md:114]: Done

File: docs/developer/example-bpf.md (hunk 5 above)
Comment (Billy99) [reply docs/developer/example-bpf.md:187]: Done

File: docs/developer/example-bpf.md [hunk 14]
```diff
@@ -0,0 +1,670 @@
...
+#### Running Unprivileged
+
+To run the `go-xdp-counter` and `go-tc-counter` examples unprivileged (without `sudo`), the
+following three steps must be performed.
+
+##### Step 1: Create `bpfd` User Group
+
+The [tutorial.md](../admin/tutorial.md) guide describes the different modes `bpfd` can be run in.
+Specifically, the [Systemd Service](../admin/tutorial.md#systemd-service) section describe how to
+start `bpfd` the `bpfd` Users and `bpfd` User Group.
```
[the remark this answers is not in the document]
Comment (Billy99) [reply docs/developer/example-bpf.md:188]: Done

File: docs/developer/example-bpf.md (hunk 6 above)
Comment (Billy99) [reply docs/developer/example-bpf.md:189]: I didn't realize it was gone. Removed documentation on it and also the const in both main.c files

File: docs/developer/example-bpf.md (hunk 7 above)
Comment (Billy99) [reply docs/developer/example-bpf.md:390]: Yes, that's my assumption.

File: docs/developer/example-bpf.md [hunk 15]
```diff
@@ -0,0 +1,670 @@
...
+The BPF bytecode should be loaded by an administrator, not the userspace program itself.
+
+![gocounter On Kubernetes](../img/gocounter-on-k8s.png)
+
+### Loading BPF Bytecode On Kubernetes
+
+This assumes there is already a Kubernetes cluster running and `bpfd` is running in the cluster
+(see [How to Manually Deploy bpfd on Kubernetes](../admin/k8s-deployment.md)).
+Instead of using the userspace program or `bpfctl` to load the BPF bytecode as done above, the bytecode
+will be loaded by creating a `BpfProgramConfig` CRD. Edit the
```
[the remark this answers is not in the document]
Comment (Billy99) [reply docs/developer/example-bpf.md:392]: Done

File: docs/developer/example-bpf.md [hunk 16]
```diff
@@ -0,0 +1,670 @@
...
+    :
+```
+
+To interact with the KubeApiServer, RBAC must be setup properly to access the `BpdProgram`
+object.
+The `bpfd-operator` defined the yaml for several ClusterRoles that can be used to access the
+different `bpfd` related CRD objects with different access rights.
+The example userspace containers will use the `bpfprogram-viewer-role`, which allows Read-Only
+access to the `BpfProgram` object.
+This ClusterRole needs to be created.
```
[the remark this answers is not in the document]
Comment (Billy99) [reply docs/developer/example-bpf.md:589]: Thanks. Removed commands to create and delete.

File: docs/developer/example-bpf.md [hunk 17]
```diff
@@ -0,0 +1,651 @@
...
+The `go-xdp-counter` and  `go-tc-counter` examples each contain a BPF Program written in C
+([xdp_counter.c](https://github.com/redhat-et/bpfd/tree/main/examples/go-xdp-counter/bpf/xdp_counter.c) and
+[tc_counter.c](https://github.com/redhat-et/bpfd/tree/main/examples/go-tc-counter/bpf/tc_counter.c))
+that is compiled into BPF bytecode.
+The BPF program counts packets that are received on an interface and stores the packet and byte
+counts in a map that is accessible by the userspace portion.
+The `go-xdp-counter` and  `go-tc-counter` examples each also have a userspace portion written in GO.
+When run locally, the userspace program makes gRPC calls to `bpfd` requesting `bpfd` to load the BPF program
+at the requested hook Point (XDP hook point or the TC hook point).
+When run in a Kubernetes deployment, the `bpfd-agen`t makes gRPC calls to `bpfd` requesting `bpfd` to load
```
Comment (astoycos) [inline docs/developer/example-bpf.md:20]: Super Nit
```suggestion
When run in a Kubernetes deployment, the `bpfd-agent` makes gRPC calls to `bpfd` requesting `bpfd` to load
```

**Review by astoycos — APPROVED** [review]
/lgtm Thanks!

**Review by anfredette — APPROVED** [review]
This looks good to me for now.

We need to figure out whether the k8s example would be better in this doc, or in the k8s deployment doc, and then probably do some combining.  Also, it might be better to go through loading and verifying both the user space and kernel space programs using our pre-built images first, and then talk about building your own container images after that.

**Pull request merged**